spring-cloud-deployer-spi/src/test/java/org/springframework/cloud/deployer/spi/app/observation/ObservedAppDeployerTests.java

```diff
@@ -0,0 +1,144 @@
+
+package org.springframework.cloud.deployer.spi.app.observation;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+
+import io.micrometer.core.instrument.Tags;
+import io.micrometer.core.tck.MeterRegistryAssert;
+import io.micrometer.tracing.Span;
+import io.micrometer.tracing.test.SampleTestRunner;
+import io.micrometer.tracing.test.simple.SpansAssert;
+import reactor.core.publisher.Mono;
+
+import org.springframework.cloud.deployer.spi.app.AppDeployer;
+import org.springframework.cloud.deployer.spi.app.AppScaleRequest;
+import org.springframework.cloud.deployer.spi.app.AppStatus;
+import org.springframework.cloud.deployer.spi.core.AppDefinition;
+import org.springframework.cloud.deployer.spi.core.AppDeploymentRequest;
+import org.springframework.cloud.deployer.spi.core.RuntimeEnvironmentInfo;
+import org.springframework.core.io.PathResource;
+
+/**
+ * Tests for {@link AppDeployer}
+ */
+public class ObservedAppDeployerTests extends SampleTestRunner {
+
+	ObservedAppDeployerTests() {
+		super(SampleRunnerConfig.builder().build());
+	}
+
+	@Override
+	public SampleTestRunnerConsumer yourCode() {
+		return (bb, meterRegistry) -> {
+			AppDeployer appDeployer = new ObservedAppDeployer(getObservationRegistry(), appDeployer(), 100L);
+
+			String id = appDeployer.deploy(deploymentRequest());
+			appDeployer.status(id);
+			appDeployer.scale(new AppScaleRequest(id, 2));
+			appDeployer.undeploy(id);
+
+			SpansAssert.assertThat(bb.getFinishedSpans())
+					.haveSameTraceId()
+					.hasASpanWithName("deploy", spanAssert -> spanAssert.hasRemoteServiceNameEqualTo("Test")
+							.hasTag("spring.cloud.deployer.app.id", "Deployment request received")
+							.hasTag("spring.cloud.deployer.platform.k8s.url", "https://foo.bar/baz")
+							.hasKindEqualTo(Span.Kind.PRODUCER))
+					.hasASpanWithName("status", spanAssert -> spanAssert.hasRemoteServiceNameEqualTo("Test")
+							.hasTag("spring.cloud.deployer.app.id", "Deployment request received")
+							.hasTag("spring.cloud.deployer.platform.k8s.url", "https://foo.bar/baz")
+							.hasKindEqualTo(Span.Kind.PRODUCER))
+					.hasASpanWithName("scale", spanAssert -> spanAssert.hasRemoteServiceNameEqualTo("Test")
+							.hasTag("spring.cloud.deployer.platform.k8s.url", "https://foo.bar/baz")
+							.hasTag("spring.cloud.deployer.scale.count", "2")
+							.hasKindEqualTo(Span.Kind.PRODUCER))
+					.hasASpanWithName("undeploy", spanAssert -> spanAssert.hasRemoteServiceNameEqualTo("Test")
+							.hasTag("spring.cloud.deployer.app.id", "Deployment request received")
+							.hasTag("spring.cloud.deployer.platform.k8s.url", "https://foo.bar/baz")
+							.hasKindEqualTo(Span.Kind.PRODUCER))
+					.hasSize(4);
+
+			MeterRegistryAssert.assertThat(getMeterRegistry())
+					.hasTimerWithNameAndTags("spring.cloud.deployer.deploy", Tags.of("error", "none", "spring.cloud.deployer.app.id", "Deployment request received", "spring.cloud.deployer.platform.k8s.url", "https://foo.bar/baz"))
```

Comment:
I would be nice to have at least 1 error case where maybe deploy/scale throws an exception.